File: src/gmeow_tools/feedback_bundle.py
# SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
# SPDX-License-Identifier: AGPL-3.0-only

"""Build, read, and verify the self-describing diagnostics feedback bundle (#654).

`gmeow-dev feedback` always emits ``dist/gmeow-feedback.gts``: a self-contained
GTS bundle whose snapshot graph IS the ``gmeow:`` RDF projection of the findings
(SPARQL-queryable), with the SARIF 2.1.0 and flat-JSON projections riding as
content-addressed blob frames. The snapshot content id is stamped into the
report metadata so the bundle is a verifiable self-attestation: re-deriving the
snapshot id from the folded graph must reproduce the stamped value.

The canonical ``gmeow.gts`` never carries a report — only this feedback bundle
does (artifact separation, no flag), per the no-optionality doctrine.
"""

from __future__ import annotations

import json
from typing import Any

import gts
from gmeow_rdf.compat.rdflib import Dataset

from gmeow_tools.gts_producer import _Builder

#: Blob representation labels for the embedded report projections.
REP_SARIF = "gmeow:report/sarif"
REP_FINDINGS = "gmeow:report/findings"  # flat JSON
#: Report-metadata key carrying the snapshot self-attestation content id.
META_SNAPSHOT_ID = "snapshotContentId"


def _findings_dataset(report: Any) -> Dataset:
    """The ``gmeow:`` RDF projection of *report* as an rdflib Dataset."""
    dataset = Dataset()
    nquads = report.to_gmeow_rdf()
    if nquads.strip():
        dataset.parse(data=nquads, format="nquads")
    return dataset


def build_feedback_bundle(report: Any) -> bytes:
    """Embed *report* into a self-describing feedback ``.gts`` bundle.

    The snapshot graph is the findings RDF; SARIF and flat JSON ride as blobs.
    The snapshot content id is stamped into the report metadata before the
    JSON/SARIF projections are rendered, so the embedded report attests to the
    bundle it lives in.
    """
    builder = _Builder()
    builder.add_graph(_findings_dataset(report))
    report.set_metadata_json(
        META_SNAPSHOT_ID, json.dumps(builder.snapshot_content_id())
    )
    sarif = report.to_sarif().encode("utf-8")
    flat = report.to_json().encode("utf-8")
    return builder.to_gts(
        report_blobs=[
            (sarif, "application/sarif+json", REP_SARIF),
            (flat, "application/json", REP_FINDINGS),
        ]
    )


def read_report_blobs(bundle: bytes) -> dict[str, bytes]:
    """Map each embedded report blob's ``rep`` to its decoded payload."""
    folded = gts.read(bundle)
    out: dict[str, bytes] = {}
    for digest, meta in folded.blob_meta.items():
        rep = meta.get("rep")
        payload = folded.blobs.get(digest)
        if rep is not None and payload is not None:
            out[str(rep)] = payload
    return out


def verify_feedback_bundle(bundle: bytes) -> bool:
    """True when the embedded report attests to this bundle's snapshot (#654).

    Re-derives the snapshot content id from the folded findings graph and checks
    it equals the ``snapshotContentId`` the embedded flat-JSON report recorded.

    The bundle is untrusted input: a verifier sits on a trust boundary, so a
    corrupt or tampered bundle (unreadable bytes, malformed JSON, a non-mapping
    payload, an unparsable graph) is simply *not a valid self-attestation* and
    returns ``False`` rather than raising — honoring the boolean contract. This
    guards external input and does not soften the no-optionality doctrine that
    governs our own pipeline's report_json.
    """
    try:
        blobs = read_report_blobs(bundle)
        flat = blobs.get(REP_FINDINGS)
        if flat is None:
            return False
        payload = json.loads(flat)
        if not isinstance(payload, dict):
            return False
        metadata = payload.get("metadata")
        if not isinstance(metadata, dict):
            return False
        stamped = metadata.get(META_SNAPSHOT_ID)
        if stamped is None:
            return False

        folded = gts.read(bundle)
        dataset = Dataset()
        nquads = gts.to_nquads(folded)
        if nquads.strip():
            dataset.parse(data=nquads, format="nquads")
        builder = _Builder()
        builder.add_graph(dataset)
        return bool(stamped == builder.snapshot_content_id())
    except Exception:
        # A bundle we cannot read, fold, or parse cannot attest to itself.
        return False

File: crates/gmeow-dev-cli/src/feedback_bundle.rs
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: AGPL-3.0-only

//! Build, read, and verify the self-describing diagnostics feedback bundle.
//!
//! `gmeow-dev feedback` always emits `dist/gmeow-feedback.gts`: a self-contained
//! GTS bundle whose snapshot graph IS the `gmeow:` RDF projection of the findings
//! (SPARQL-queryable), with the SARIF 2.1.0 and flat-JSON projections riding as
//! content-addressed blob frames. The snapshot content id is stamped into the
//! report metadata so the bundle is a verifiable self-attestation.
//!
//! This module is the Rust twin of the retired Python
//! `src/gmeow_tools/feedback_bundle.py` + `src/gmeow_tools/gts_producer.py`.

use std::collections::BTreeMap;

use ciborium::value::Value as CborValue;
use gmeow_errors::{Report, render};
use purrdf::gts::dataset_from_gts_graph;
use purrdf::gts::reader::read;
use purrdf::gts_compose::{BlobRow, SnapshotBuilder};
use purrdf::parse_dataset;
use serde_json::Value;

/// Blob representation label for the embedded SARIF 2.1.0 projection.
pub const REP_SARIF: &str = "gmeow:report/sarif";
/// Blob representation label for the embedded flat-JSON findings projection.
pub const REP_FINDINGS: &str = "gmeow:report/findings";
/// Report-metadata key carrying the snapshot self-attestation content id.
pub const META_SNAPSHOT_ID: &str = "snapshotContentId";

/// Embed `report` into a self-describing feedback `.gts` bundle.
///
/// The snapshot graph is the findings RDF; SARIF and flat JSON ride as blobs.
/// The snapshot content id is stamped into the report metadata before the
/// JSON/SARIF projections are rendered, so the embedded report attests to the
/// bundle it lives in.
pub fn build_feedback_bundle(report: &Report) -> gmeow_errors::Result<Vec<u8>> {
    let mut builder = SnapshotBuilder::new();
    let nquads = render::to_gmeow_rdf(report);
    if !nquads.trim().is_empty() {
        let dataset = parse_dataset(nquads.as_bytes(), "application/n-quads", None)
            .map_err(|e| crate::error::feedback(format!("parse findings RDF: {e}")))?;
        builder
            .add_dataset(&dataset)
            .map_err(|e| crate::error::feedback(format!("add findings dataset: {e}")))?;
    }

    let snapshot_id = builder.snapshot_content_id();

    let mut stamped = report.clone();
    stamped
        .metadata
        .insert(META_SNAPSHOT_ID.to_owned(), Value::String(snapshot_id));

    let sarif = render::to_sarif(&stamped)
        .map_err(|e| crate::error::feedback(format!("sarif render: {e}")))?;
    let flat = render::to_json(&stamped)
        .map_err(|e| crate::error::feedback(format!("json render: {e}")))?;

    gmeow_gts_profile::emit_gmeow_gts(
        &builder,
        Vec::new(),
        vec![
            BlobRow {
                data: sarif.into_bytes(),
                media_type: "application/sarif+json".to_owned(),
                rep: REP_SARIF.to_owned(),
            },
            BlobRow {
                data: flat.into_bytes(),
                media_type: "application/json".to_owned(),
                rep: REP_FINDINGS.to_owned(),
            },
        ],
        None,
        None,
        None,
    )
    .map_err(|e| crate::error::feedback(format!("emit feedback bundle: {e}")))
}

/// Map each embedded report blob's `rep` to its decoded payload.
pub fn read_report_blobs(
    graph: &mut purrdf::gts::model::Graph,
) -> gmeow_errors::Result<BTreeMap<String, Vec<u8>>> {
    let decoded = graph
        .decoded_blobs()
        .map_err(|e| crate::error::feedback(format!("decode blobs: {e}")))?;
    let by_digest: BTreeMap<&str, &[u8]> = decoded
        .iter()
        .map(|(digest, bytes)| (digest.as_str(), bytes.as_slice()))
        .collect();

    let mut out = BTreeMap::new();
    for (digest, meta) in &graph.blob_meta {
        if let Some(rep) = meta_text(meta, "rep")
            && let Some(&bytes) = by_digest.get(digest.as_str())
        {
            out.insert(rep, bytes.to_vec());
        }
    }
    Ok(out)
}

/// True when the embedded report attests to this bundle's snapshot.
///
/// Re-derives the snapshot content id from the folded findings graph and checks
/// it equals the `snapshotContentId` the embedded flat-JSON report recorded.
///
/// The bundle is untrusted input: a corrupt or tampered bundle (unreadable bytes,
/// malformed JSON, a non-mapping payload, an unparsable graph) is simply *not a
/// valid self-attestation* and returns `False` rather than raising.
pub fn verify_feedback_bundle(bundle: &[u8]) -> bool {
    let mut graph = read(bundle, true, None);

    let Some(findings_digest) = graph.blob_meta.iter().find_map(|(digest, meta)| {
        meta_text(meta, "rep")
            .filter(|rep| rep == REP_FINDINGS)
            .map(|_| digest.clone())
    }) else {
        return false;
    };

    let Ok(Some(flat)) = graph.blob_bytes_cloned(&findings_digest) else {
        return false;
    };

    let Ok(payload) = serde_json::from_slice::<Value>(&flat) else {
        return false;
    };

    let Some(metadata) = payload.get("metadata").and_then(Value::as_object) else {
        return false;
    };

    let Some(stamped) = metadata.get(META_SNAPSHOT_ID).and_then(Value::as_str) else {
        return false;
    };

    let Ok(dataset) = dataset_from_gts_graph(&graph) else {
        return false;
    };

    let mut builder = SnapshotBuilder::new();
    if builder.add_dataset(dataset.as_ref()).is_err() {
        return false;
    }

    builder.snapshot_content_id() == stamped
}

/// Extract a text value from a CBOR map's public metadata.
fn meta_text(meta: &CborValue, key: &str) -> Option<String> {
    let CborValue::Map(entries) = meta else {
        return None;
    };
    entries.iter().find_map(|(k, v)| match (k, v) {
        (CborValue::Text(k), CborValue::Text(v)) if k == key => Some(v.clone()),
        _ => None,
    })
}
